src/os/win/pa_win_version.c

```diff
@@ -0,0 +1,240 @@
+ * they can be incorporated into the canonical version. It is also
+ * requested that these non-binding requests be included along with the
+ * license above.
+ */
+
+/** @file
+ @ingroup win_src
+
+ @brief Portable implementation of Windows OS version getter.
+*/
+
+#include <windows.h>
+
+#include "pa_win_version.h"
+#include "pa_debugprint.h"
+
+// WinRT
+#if defined(WINAPI_FAMILY) && (WINAPI_FAMILY == WINAPI_FAMILY_APP)
+    #define PA_WINRT
+#endif
+
+// Alternative way for checking Windows version, allows to check version of Windows 8.1 and up
+#ifndef PA_WINRT
+static BOOL IsWindowsVersionOrGreater(WORD wMajorVersion, WORD wMinorVersion, WORD wServicePackMajor)
+{
+    typedef ULONGLONG (NTAPI *LPFN_VERSETCONDITIONMASK)(ULONGLONG ConditionMask, DWORD TypeMask, BYTE Condition);
+    typedef BOOL (WINAPI *LPFN_VERIFYVERSIONINFO)(LPOSVERSIONINFOEXA lpVersionInformation, DWORD dwTypeMask, DWORDLONG dwlConditionMask);
+
+    LPFN_VERSETCONDITIONMASK fnVerSetConditionMask;
+    LPFN_VERIFYVERSIONINFO fnVerifyVersionInfo;
+    OSVERSIONINFOEXA osvi = { sizeof(osvi), 0, 0, 0, 0, {0}, 0, 0 };
+    DWORDLONG dwlConditionMask;
+
+    fnVerSetConditionMask = (LPFN_VERSETCONDITIONMASK)GetProcAddress(GetModuleHandleA("kernel32"), "VerSetConditionMask");
+    fnVerifyVersionInfo = (LPFN_VERIFYVERSIONINFO)GetProcAddress(GetModuleHandleA("kernel32"), "VerifyVersionInfoA");
+
+    if ((fnVerSetConditionMask == NULL) || (fnVerifyVersionInfo == NULL))
+        return FALSE;
+
+    dwlConditionMask = fnVerSetConditionMask(
+        fnVerSetConditionMask(
+            fnVerSetConditionMask(
+                0, VER_MAJORVERSION,     VER_GREATER_EQUAL),
+                   VER_MINORVERSION,     VER_GREATER_EQUAL),
+                   VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL);
+
+    osvi.dwMajorVersion    = wMajorVersion;
+    osvi.dwMinorVersion    = wMinorVersion;
+    osvi.wServicePackMajor = wServicePackMajor;
+
+    return (fnVerifyVersionInfo(&osvi, VER_MAJORVERSION | VER_MINORVERSION | VER_SERVICEPACKMAJOR, dwlConditionMask) != FALSE);
+}
+#endif
+
+static PaOsVersion GetOsVersion()
+{
+    PaOsVersion version = paOsVersionWindowsUnknown;
+
+#ifndef PA_WINRT
+    DWORD dwMajorVersion = 0xFFFFFFFFU, dwMinorVersion = 0, dwPlatformId = 0, dwBuild = 0;
+
+    // Can be missing in some MinGW distributions
+#ifndef NT_SUCCESS
+    typedef LONG NTSTATUS;
+#endif
+
+#ifndef VER_PLATFORM_WIN32_NT
+    #define VER_PLATFORM_WIN32_NT 2
+#endif
+
+    // RTL_OSVERSIONINFOW equals OSVERSIONINFOW but it is missing inb MinGW winnt.h header,
+    // thus use OSVERSIONINFOW for greater portability
+    typedef NTSTATUS (WINAPI *LPFN_RTLGETVERSION)(POSVERSIONINFOW lpVersionInformation);
+    LPFN_RTLGETVERSION fnRtlGetVersion;
+
+    #define NTSTATUS_SUCCESS ((NTSTATUS)0x00000000L)
+
+    // RtlGetVersion must be able to provide true Windows version (Windows 10 may be reported as Windows 8
+    // by GetVersion API)
```

Comment:
`RtlGetVersion` does in fact also lie about the version in some cases, for example when Windows decides to use compatibility shims.

Reply:
@MichalPetryka thank you for your comment, I amended inline comment accordingly.